sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,187 @@
+package queue
+
+import (
+	"container/heap"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/mitchellh/copystructure"
+)
+
+// ErrEmpty is returned for queues with no items
+var ErrEmpty = errors.New("queue is empty")
+
+// ErrDuplicateItem is returned when the queue attmepts to push an item to a key that
+// already exists. The queue does not attempt to update, instead returns this
+// error. If an Item needs to be updated or replaced, pop the item first.
+var ErrDuplicateItem = errors.New("duplicate item")
+
+// ErrItemNotFound is a struct that implements the error interface
+var _ error = &ErrItemNotFound{}
+
+type ErrItemNotFound struct {
+	Key string
+}
+
+func (e *ErrItemNotFound) Error() string {
+	return fmt.Sprintf("queue item with key (%s) not found", e.Key)
+}
+
+// NewErrItemNotFound creates a "not found" error for the given key
+func NewErrItemNotFound(key string) *ErrItemNotFound {
+	return &ErrItemNotFound{
+		Key: key,
+	}
+}
+
+// Item is something managed in the priority queue
+type Item struct {
+	// Key is a unique string used to identify items in the internal data map
+	Key string
+	// Value is an unspecified type that implementations can use to store
+	// information
+	Value interface{}
+
+	// Priority determines ordering in the queue, with the lowest value being the
+	// highest priority
+	Priority int64
```

Comment:
Why `int64` instead of `int` for the priority?

Reply:
Right now the coming implementation that uses the Queue is time based, and we're using `<time>.Unix()` to return the time since unix epoch, so it made sense to just keep at `int64`.